### PR TITLE
Add missing tox env checks to Github checks (and fix failures)

### PR DESCRIPTION
**Pull request by gaborbernat**
Add the lint tox environment as Github action test via https://github.com/pre-commit/action

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -14,13 +14,24 @@ repos:
   - id: debug-statements
   - id: forbid-new-submodules
 - repo: https://github.com/psf/black
-  rev: stable
+  rev: 19.10b0
```
Comment (gaborbernat) [inline .pre-commit-config.yaml:17]: it's better to have exact versions, as a new version doesn't break black.

Instead you can use ``pre-commit autoupgrade`` to autobump lint versions periodically.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -14,13 +14,24 @@ repos:
...
   hooks:
   - id: black
 - repo: https://gitlab.com/pycqa/flake8
-  rev: 3.8.1
+  rev: "3.8.1"
   hooks:
   - id: flake8
+    additional_dependencies: ["flake8-bugbear == 20.1.2"]
+    language_version: python3.8
+- repo: https://github.com/asottile/seed-isort-config
```
Comment (gaborbernat) [inline .pre-commit-config.yaml:26]: instead manually populating known_third_party in setup.cfg use this tool to autogenerate

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -14,13 +14,24 @@ repos:
...
+  rev: v2.1.1
+  hooks:
+  - id: seed-isort-config
+    args: [--application-directories, '.:src']
+- repo: https://github.com/asottile/setup-cfg-fmt
```
Comment (gaborbernat) [inline .pre-commit-config.yaml:31]: add formatter for setup.cfg :-) similar as black is for python files. Bonus it automatically updates the python classifiers based on python requires, to ensure they are always in sync

File: src/editables/__init__.py [hunk 4]
```diff
@@ -1,5 +1,4 @@
 import pkgutil
-import os
```
Comment (gaborbernat) [inline src/editables/__init__.py:2]: the linter failed as this import is not used

File: src/editables/install_hook.py [hunk 5]
```diff
@@ -31,4 +31,5 @@ def _bootstrap():
     sys.modules[__name__] = module
     spec.loader.exec_module(module)
 
+
```
Comment (gaborbernat) [inline src/editables/install_hook.py:34]: and here needed to be a new line

File: .pre-commit-config.yaml [hunk 6]
```diff
@@ -14,13 +14,24 @@ repos:
   - id: debug-statements
   - id: forbid-new-submodules
 - repo: https://github.com/psf/black
-  rev: stable
+  rev: 19.10b0
   hooks:
   - id: black
 - repo: https://gitlab.com/pycqa/flake8
-  rev: 3.8.1
+  rev: "3.8.1"
```
Comment (pfmoore) [inline .pre-commit-config.yaml:21]: Why is this one in quotes when the others aren't? I find YAML syntax fiddly enough that I'd prefer consistency.

File: setup.cfg [hunk 7]
```diff
@@ -6,9 +6,6 @@ long_description = file: README.md
 long_description_content_type = text/markdown
 url = https://github.com/pfmoore/editables
 author = Paul Moore
-author-email = p.f.moore@gmail.com
-maintainer = Paul Moore
-maintainer-email = p.f.moore@gmail.com
```
Comment (pfmoore) [inline setup.cfg:11]: Did the auto-formatter do this? I'm not keen, I'd rather keep this data together so that if any edits are needed it's easy to see that they are needed in multiple places. I'll take a look at the auto-formatter and see what config options it has.

File: .pre-commit-config.yaml (hunk 3 above)
Comment (pfmoore) [reply .pre-commit-config.yaml:31]: I'll take a look at this and see what I think (see below).

File: .github/workflows/pre-commit.yml [hunk 8]
```diff
@@ -0,0 +1,21 @@
+# per https://github.com/pre-commit/action
+name: pre-commit
```
Comment (pfmoore) [inline .github/workflows/pre-commit.yml:2]: Thanks, this was on my TODO list to add.

File: .pre-commit-config.yaml (hunk 6 above)
Comment (gaborbernat) [reply .pre-commit-config.yaml:21]: Because YAML is picky... and dumb at the same time. If something starts with numbers and then a ``.`` it thinks it's a floating number. Others start with ``v`` to avoid this issue.

File: setup.cfg (hunk 7 above)
Comment (gaborbernat) [reply setup.cfg:11]: Yes; @asottile might know what's the reason for the order not being kept.

**Comment by gaborbernat** [comment]
@pfmoore I'm done now. All checks run and pass. I'll do the coverage tracking for the install hook once this gets in.

File: .pre-commit-config.yaml (hunk 2 above)
Comment (pfmoore) [reply .pre-commit-config.yaml:26]: Is that what added setuptools? That seems pointless, as it's only used in `setup.py`, but I guess it's harmless.

File: .pre-commit-config.yaml (hunk 6 above)
Comment (pfmoore) [reply .pre-commit-config.yaml:21]: OK, so let's quote them all for consistency, I guess :-(

But don't do that yet, I'll check what other projects do and come to a decision on what style I like. All options seem to suck, unfortunately.

File: .pre-commit-config.yaml (hunk 2 above)
Comment (gaborbernat) [reply .pre-commit-config.yaml:26]: Well, the list here is for all python files in the projects. And setup.py is also a python file within this project. E.g. if you would have a Sphinx ``conf.py`` that pulls in sphinx modules via import that would also show up here. But it's harmless as you said.

File: .github/workflows/tests.yml [hunk 9]
```diff
@@ -16,10 +16,10 @@ jobs:
         - Windows
         - MacOS
         py:
-        - 3.5
-        - 3.6
-        - 3.7
         - 3.8
+        - 3.7
+        - 3.6
+        - 3.5
```
Comment (pfmoore) [inline .github/workflows/tests.yml:22]: Minor point, but this is unrelated so I would have preferred a separate PR.

File: .pre-commit-config.yaml (hunk 6 above)
Comment (gaborbernat) [reply .pre-commit-config.yaml:21]: Well, I did not change it. Rather ``pre-commit autoupgrade`` did. The ``pre-commit`` policy is to use the non-quoted version where it's needed to remove ambiguity (as is here), but keep the short version otherwise where it's needed as it's more readable. IMHO this is a good enough policy to me. This file, in general, is not meant to be human maintained, but rather by ``pre-commit`` itself once the checks have been added. I don't think ``pre-commit`` has an ``always-quote`` option, @asottile?

File: .github/workflows/tests.yml (hunk 9 above)
Comment (gaborbernat) [reply .github/workflows/tests.yml]: Removed 🤷

File: .github/workflows/other.yml [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+    steps:
+    - uses: actions/checkout@v2
+    - name: setup Python 3.8
+      uses: actions/setup-python@v1
+      with:
+        python-version: 3.8
+    - name: install tox
+      run: python -m pip install tox
+    - name: run check for ${{ matrix.tox_env }}
+      run: python -m tox -e ${{ matrix.tox_env }}
```
Comment (pfmoore) [inline .github/workflows/other.yml:25]: I'm getting confused. Can we not have this as part of the "lint" checks, I feel it fits better in there.

I'd be OK with merging this and fixing that later, but it'll have to wait on the question about the setup.cfg formatter.

Also, I'm finding that the github UI is getting very confusing here. You seem to have pushed extra commits, but I only saw them by carefully checking what had been added. I don't know if that was related to the fact that you force-pushed, or if it's because you added new commits with extra changes in different files, but if you could do anything to simplify the review process, I'd appreciate it. (In particular, while you're still working on a PR, please try to avoid force-pushes and history edits - once the PR looks OK, if you want to tidy up the commit history then, let's do it as a single pass, although personally I'm fine with not doing that and just merging with the actual commit history as developed).

File: .pre-commit-config.yaml (hunk 6 above)
Comment (pfmoore) [reply .pre-commit-config.yaml:21]: OK, so maybe we don't bother with that helper for now, we can do things manually in the short term. If checking what the tool did is more effort than not using the tool, I'm fine with holding off on using the tool for now.

File: .github/workflows/other.yml (hunk 10 above)
Comment (gaborbernat) [reply .github/workflows/other.yml:25]: > I'm getting confused. Can we not have this as part of the "lint" checks, I feel it fits better in there.

Well, this is a check on the binary packages generated (sdist and wheel) so they don't link check the existing code (that pre-commit does). So I don't think it can fall under that umbrella, and there's no way to force put it under AFAIK (unless we write a pre-commit plugin for it, but again as it's not a lint check I don't think it should be there).

File: .pre-commit-config.yaml (hunk 6 above)
Comment (gaborbernat) [reply .pre-commit-config.yaml:21]: Why do you care so strongly about how a tool generates a configuration files content as long as that is done automatically?

Comment (pfmoore) [reply .pre-commit-config.yaml:21]: Maybe I don't :-) I'll have a think - I'm used to manually editing at least some of this stuff, so maybe I need to just embrace a new approach :-)

File: setup.cfg (hunk 7 above)
Comment (asottile) [reply setup.cfg:11]: ah, it moved these because it doesn't know about the dashed spelling -- the canonical spelling for setuptools is `author_email` and `maintainer_email`

File: .pre-commit-config.yaml (hunk 6 above)
Comment (asottile) [reply .pre-commit-config.yaml:21]: huh, it shouldn't have quoted this one -- must be the feature that someone just added let me check that 🤔

Comment (asottile) [reply .pre-commit-config.yaml:21]: doesn't quote it for me 🤷  -- but as of pre-commit 2.4 if you quote them then pre-commit will keep them quoted after autoupdate

**Comment by gaborbernat** [comment]
@pfmoore addressed your requests, is now ok?

**Comment by pfmoore** [comment]
I'm going to accept these changes in the interests of moving forwards. However, at this point I want to stop looking at the project infrastructure and focus on the actual code, so for the immediate future I won't be accepting any more project infrastructure changes. I may make infrastructure changes myself (one I'm aware of is the coverage of the template) but I'd prefer to do those myself rather than accept PRs for them.
